### PR TITLE
Partial/full activation mode and multi root. Disable automatic configures.

### PR DESCRIPTION
**Pull request by andreeis**
Introduce cmake.configureOnEdit to control the automatic configures triggered by editing CMakeLists.txt or cmake.sourceDirectory.

Refactored and fixed the whole partial/full activation modes in the multi root scenario.

This addresses GitHub issues:
https://github.com/microsoft/vscode-cmake-tools/issues/1259
https://github.com/microsoft/vscode-cmake-tools/issues/1269

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1249,6 +1249,12 @@
           "description": "%cmake-tools.configuration.cmake.configureOnOpen.description%",
           "scope": "resource"
         },
+        "cmake.configureOnEdit": {
+          "type": "boolean",
+          "default": true,
+          "description": "%cmake-tools.configuration.cmake.configureOnOpen.description%",
```
Comment (bobbrow) [inline package.json:1255]: typo: `configureOnEdit`

File: src/cmake-tools.ts [hunk 2]
```diff
@@ -271,8 +271,15 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
           const cmakeListsFile = await vscode.window.showOpenDialog(openOpts);
           if (cmakeListsFile) {
             const fullPathDir: string = path.parse(cmakeListsFile[0].fsPath).dir;
-            const relPathDir: string = path.relative(this.folder.uri.fsPath, fullPathDir);
-            vscode.workspace.getConfiguration('cmake').update("sourceDirectory", path.join("${workspaceFolder}", relPathDir));
+            const relPathDir: string = lightNormalizePath(path.relative(this.folder.uri.fsPath, fullPathDir));
```
Comment (bobbrow) [inline src/cmake-tools.ts:274]: Does this give you `../folder` for folders outside of `${workspaceFolder}`?

File: src/cmake-tools.ts [hunk 3]
```diff
@@ -271,8 +271,15 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
+            const drv = await this.getCMakeDriverInstance();
+            if (drv) {
+              const joinedPath = "${workspaceFolder}/".concat(relPathDir);
```
Comment (bobbrow) [inline src/cmake-tools.ts:277]: `path.join` is probably better to use than `.concat`

File: src/cmake-tools.ts [hunk 4]
```diff
@@ -271,8 +271,15 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
+              vscode.workspace.getConfiguration('cmake', this.folder.uri).update("sourceDirectory", joinedPath);
+              drv.config.updatePartial({sourceDirectory: joinedPath});
+            } else {
+              throw new Error(localize('unable.to.save.sourceDirectory', 'Unable to save "cmake.sourceDirectory". There is no valid cmake driver instance.'));
```
Comment (bobbrow) [inline src/cmake-tools.ts:281]: You don't need a `drv` to set `cmake.sourceDirectory`, so I think you should set it unconditionally. In that case you could just instruct the user to reload the window and the setting should take effect.  E.g. "Something went wrong while setting the cmake.sourceDirectory setting. Please run the "Reload Window" command for the change to take effect") Or give them a button to run the command automatically.

File: src/cmake-tools.ts [hunk 5]
```diff
@@ -445,14 +452,22 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
       const str = td.uri.fsPath;
-      if (str.endsWith("CMakeLists.txt")) {
-        // Reset to a full activation mode.
-        // If required, the extension will switch back to partial activation mode
-        // after more analysis triggered by the below configure.
-        await enableFullFeatureSet(true);
-
-        log.debug(localize('cmakelists.save.trigger.reconfigure', "We are saving a CMakeLists.txt file, attempting automatic reconfigure..."));
-        await this.configure([], ConfigureType.Normal);
+      if (str === path.join(this.folder.uri.fsPath, "CMakeLists.txt")) {
```
Comment (bobbrow) [inline src/cmake-tools.ts:455]: Is this a multiroot problem that you're trying to fix?  If so, then I think you need to look at the `cmake.sourceDirectory` setting and look if the CMakeLists.txt is under that folder instead. 

Technically, I think that CMakeLists can reference folders anywhere in the filesystem, so that's not a 100% fix, but maybe addresses 99% of cases.

File: src/cmake-tools.ts [hunk 6]
```diff
@@ -445,14 +452,22 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
+        // should be full or partial, so there is no need to proactively enable full here,
+        // unless the automatic configure is disabled.
+        // If there is a configure or a build in progress, we should avoid setting full activation here,
+        // even if cmake.configureOnEdit is true, because this may overwrite a different decision
+        // that was done earlier by that ongoing configure process.
+        const drv = await this.getCMakeDriverInstance();
+        if (drv && !drv.configOrBuildInProgress()) {
+          if (drv.config.configureOnEdit) {
+            log.debug(localize('cmakelists.save.trigger.reconfigure', "Detected saving of CMakeLists.txt, attempting automatic reconfigure..."));
+            await this.configure([], ConfigureType.Normal);
```
Comment (bobbrow) [inline src/cmake-tools.ts:466]: Does `configure` enable the feature set?

File: src/cmake-tools.ts [hunk 7]
```diff
@@ -445,14 +452,22 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
+          } else {
+            await enableFullFeatureSet(true, this.folder);
+          }
+        }
```
Comment (bobbrow) [inline src/cmake-tools.ts:470]: I think for the else of `configOrBuildInProgress` you should log a message that the user needs to reconfigure and build again.

File: src/drivers/driver.ts [hunk 8]
```diff
@@ -634,6 +639,15 @@ export abstract class CMakeDriver implements vscode.Disposable {
         return -1;
       }
 
+      // _beforeConfigureOrBuild contains the only scenario when we would want partial features set
```
Comment (bobbrow) [inline src/drivers/driver.ts:642]: If this is the case, then should the call to enable the full feature set go there instead? This would also give you the opportunity to turn it off if the conditions started to fail for some reason.

File: src/drivers/cmfileapi-driver.ts [hunk 9]
```diff
@@ -97,6 +102,26 @@ export class CMakeFileApiDriver extends codemodel.CodeModelDriver {
       log.debug(`Reload CMake cache: ${this.cachePath} changed`);
       rollbar.invokeAsync('Reloading CMake Cache', () => this.updateCodeModel());
     });
+
+    this.config.onChange('sourceDirectory', async () => {
```
Comment (bobbrow) [inline src/drivers/cmfileapi-driver.ts:106]: Should this config listener be enabled for all drivers (not just file api)?

File: src/drivers/cmfileapi-driver.ts [hunk 10]
```diff
@@ -97,6 +102,26 @@ export class CMakeFileApiDriver extends codemodel.CodeModelDriver {
...
+      // unless the automatic configure is disabled.
+      // If there is a configure or a build in progress, we should avoid setting full activation here,
+      // even if cmake.configureOnEdit is true, because this may overwrite a different decision
+      // that was done earlier by that ongoing configure process.
+      if (!this.configOrBuildInProgress()) {
+        if (this.config.configureOnEdit) {
+          log.debug(localize('cmakelists.save.trigger.reconfigure', "Detected 'cmake.sourceDirectory' setting update, attempting automatic reconfigure..."));
+          await this.configure([]);
+        } else if (this.workspaceFolder) {
+          const folder = vscode.workspace.getWorkspaceFolder(vscode.Uri.file((this.workspaceFolder)));
```
Comment (bobbrow) [inline src/drivers/cmfileapi-driver.ts:118]: extra parenthesis in the function call

File: src/extension.ts [hunk 11]
```diff
@@ -182,30 +174,24 @@ class ExtensionManager implements vscode.Disposable {
...
+  // The context variable "cmake:enableFullFeatureSet" (which controls
+  // all the cmake commands and UI elements) is set to true,
+  // if there is at least one folder with full features set enabled.
+  // We need to add this private _enableFullFeaturesSetOnWorkspace here
+  // because currently there is no way of reading a context variable
+  // like cmake:enableFullFeatureSet, to apply the OR operation on it.
+  private _enableFullFeatureSetOnWorkspace = false;
+  public enableFullFeatureSet(fullFeatureSet: boolean, folder: vscode.WorkspaceFolder) {
+    this.getFolderContext(folder).ignoreCMakeListsMissing = !fullFeatureSet;
+    this._enableFullFeatureSetOnWorkspace = this._enableFullFeatureSetOnWorkspace || fullFeatureSet;
```
Comment (bobbrow) [inline src/extension.ts:192]: I find this pattern confusing. You need to set `_enableFullFeatureSetOnWorkspace` to `false` outside of this method because `fullFeatureSet` can't change the true to false. Is there a reason you can't just take the value of `fullFeatureSet`?

File: src/cmake-tools.ts (hunk 5 above)
Comment (bobbrow) [reply src/cmake-tools.ts]: We also support `sourceDirectory` with `CMakeLists.txt` included in the path so you should account for that here.

File: src/cmake-tools.ts [hunk 12]
```diff
@@ -445,14 +458,27 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
...
+        if (drv && !drv.configOrBuildInProgress()) {
+          if (drv.config.configureOnEdit) {
+            log.debug(localize('cmakelists.save.trigger.reconfigure', "Detected saving of CMakeLists.txt, attempting automatic reconfigure..."));
+            await this.configure([], ConfigureType.Normal);
+          } else {
+            await enableFullFeatureSet(true, this.folder);
+          }
+        } else {
+          log.warning(localize('cmakelists.save.could.not.reconfigure',
+           'Detected saving of CMakeLists.txt but could not configure because there is no valid driver or the project is already building/configuring.'));
```
Comment (bobbrow) [inline src/cmake-tools.ts:479]: >Detected saving of CMakeLists.tx [](start = 12, length = 32)

I would reword it like this:

"Changes were detected in CMakeLists.txt but we could not reconfigure the project because another operation is already in progress."

File: src/cmake-tools.ts (hunk 5 above)
Comment (andreeis) [reply src/cmake-tools.ts]: Good catch. It made me remember https://github.com/microsoft/vscode-cmake-tools/pull/1016/files and I actually debugged and I confirm that CMakeLists.txt gets removed from sourceDirectory before the CMakeLists path check. Do you think there is a different scenario where actually sourceDirectory would end with CMakeLists.txt once we hit the file content change listener?

**Comment by andreeis** [comment]
I am looking at the failed tests. Again, "electron: Loading non context-aware native modules in the renderer process is deprecated...". I fixed this before with another PR a few weeks or months ago. I had to move a particular call somewhere else.

**Comment by bobbrow** [comment]
>     return cache.get('CMAKE_GENERATOR')!.value;

can you use `?.` and make the return type optional?  I think you only compare this value to an optional.

(I like to avoid using `!.` where ever possible)

---
Refers to: src/drivers/cmfileapi-driver.ts:72 in 27bee85. [](commit_id = 27bee8554bc2ea229f5c31ab01b3e91252536d29, deletion_comment = False)

**Comment by bobbrow** [comment]
>     return cache.get('CMAKE_GENERATOR')!.value;

Oh, it looks like this was merged in from another PR.  You can disregard.

---
In reply to: [642980937](https://github.com/Microsoft/vscode-cmake-tools/pull/1295#issuecomment-642980937) [](ancestors = 642980937)

---
Refers to: src/drivers/cmfileapi-driver.ts:72 in 159e0e6. [](commit_id = 159e0e6886bc9880c35715267ed656b78a623e6d, deletion_comment = False)

File: src/extension.ts (hunk 11 above)
Comment (bobbrow) [reply src/extension.ts:192]: Do we ever need to turn support off?  I'm wondering if we should just leave the full feature set enabled once it's been enabled.

Comment (bobbrow) [reply src/extension.ts:192]: If you remove the state that would make CMake Tools enable the full feature set, then you would need to reload the window at that point to disable it.  I would be fine with that behavior.

**Comment by andreeis** [comment]
The only fix needed for the Electron issue was to not call enableFullFeatureSet from driver.ts. To still achieve the same functionality (we want full activation when beforeConfigureOrBuild returns true and partial only for missing CMakeLists.txt), I introduced a -2 error code (which wasn't used already, only -1 was returned on any failure path) to recognize later that we have a missing CMakeLists.txt, thus calling enableFullFeatureSet in a place where Electron doesn't complain.

The reason for commenting out the sourceDirectory listeners in the driver (the previous commit) was that I had infrastructure issues on my Ubuntu and with no amount of cleanup (not even fresh cloning of CMake Tools) I wasn't able to get rid of the Electron test issue so I just made some experiments to leverage the GitHub CI.
